draft-ietf-httpbis-resumable-upload.md

```diff
@@ -220,9 +220,14 @@ If the request completes successfully and the entire upload is complete, the ser
 
 If the request completes successfully but the entire upload is not yet complete, as indicated by an `Upload-Complete` field value of false in the request, the server MUST acknowledge it by responding with the `201 (Created)` status code and an `Upload-Complete` header value set to false.
 
-If the request includes an `Upload-Complete` field value set to true and a valid `Content-Length` header field, the client attempts to upload a fixed-length resource in one request. In this case, the upload's final size is the `Content-Length` field value and the server MUST record it to ensure its consistency.
+The request can indicate the upload's final size in two different ways. Both indicators may be present in the same request as long as they convey the same size. If the sizes are inconsistent, the server MUST reject the request by responding with a `400 (Bad Request)` status code.
 
-The server MAY enforce a maximum size of an upload resource. This limit MAY be equal to the upload's final size, if `Upload-Complete: ?1` and `Content-Length` are present in the upload creation request, or an arbitrary value. The limit's value or its existence MUST NOT change throughout the lifetime of the upload resource. The server MAY indicate such a limit to the client by including the `Upload-Limit` header field in the informational or final response to upload creation. If the client receives an `Upload-Limit` header field indicating that the maximum size is less than the amount of bytes it intends to upload to a resource, it SHOULD stop the current upload transfer immediately and cancel the upload ({{upload-cancellation}}).
+- If the request includes an `Upload-Complete` field value set to true and a valid `Content-Length` header field, the client attempts to upload a fixed-length resource in one request. In this case, the upload's final size is the `Content-Length` field value and the server MUST record it to ensure its consistency. The value cannot be used if the `Content-Length` header is missing (e.g. because of HTTP/1.1 chunked transfer coding, see {{Section 7.1 of RFC9112}}) or if the upload is split across multiple requests.
+- If the request includes the `Upload-Length` header field, the server MUST record its value as the upload's final size. A client SHOULD provide this header field if the upload length is known at the time of upload creation.
+
+The upload is not automatically completed if the offset reaches the upload's final size. Instead, a client MUST indicate the completion of an upload through the `Upload-Complete` header field. Indicating an upload's final size can help the server allocate necessary resources for the upload and provide early feedback if the size does not match the server's limits ({{upload-limit}}).
```

Comment:
This means that the client could upload the full resource but not complete it, then use a 0 length append to complete the upload? That would be a good test case to write for server implementations

Reply:
Yes, this is possible and was possible before. Nevertheless, I agree that this is an easy edge case to miss when implementing a server.

Reply:
I agree that this is a point where implementations have to be careful, but we also wanted to avoid using Upload-Length to decide whether an upload is complete. If so, there would be two values that can influence the completeness of an upload. And that seemed to be the worse evil than allow uploads to be completely full but not completed.

Reply:
Yeah agreed. It's good enough as it is now.